File: ugsStatusPublisher-server/src/main/java/no/tjer/teamcity/ugsStatusPublisher/UgsPublisherSettings.java
/*
 * Copyright 2000-2022 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package no.tjer.teamcity.ugsStatusPublisher;

import java.util.*;

import jetbrains.buildServer.commitPublisher.*;
import jetbrains.buildServer.commitPublisher.CommitStatusPublisher.Event;
import jetbrains.buildServer.serverSide.*;
import jetbrains.buildServer.serverSide.auth.SecurityContext;
import jetbrains.buildServer.util.StringUtil;
import jetbrains.buildServer.util.ssl.SSLTrustStoreProvider;
import jetbrains.buildServer.vcs.VcsRoot;
import jetbrains.buildServer.web.openapi.PluginDescriptor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;


/**
 * Settings for Unreal Game Sync commit status publisher.
 */
public class UgsPublisherSettings extends BasePublisherSettings implements CommitStatusPublisherSettings {

    private final SecurityContext mySecurityContext;
    private static final Set<Event> mySupportedEvents = new HashSet<Event>() {{
        add(Event.STARTED);
        add(Event.FINISHED);
        add(Event.INTERRUPTED);
        add(Event.MARKED_AS_SUCCESSFUL);
    }};

    private static final Set<Event> mySupportedEventsWithQueued = new HashSet<Event>() {{
        add(Event.QUEUED);
        add(Event.REMOVED_FROM_QUEUE);
        addAll(mySupportedEvents);
    }};

    public UgsPublisherSettings(@NotNull PluginDescriptor descriptor,
                                @NotNull WebLinks links,
                                @NotNull CommitStatusPublisherProblems problems,
                                @NotNull SecurityContext securityContext,
                                @NotNull SSLTrustStoreProvider trustStoreProvider) {
        super(descriptor, links, problems, trustStoreProvider);
        mySecurityContext = securityContext;
    }

    @NotNull
    public String getId() {
        return UgsConstants.ID;
    }

    @NotNull
    public String getName() {
        return "Unreal Game Sync";
    }

    @Nullable
    public String getEditSettingsUrl() {
        return myDescriptor.getPluginResourcesPath("ugsSettings.jsp");
    }

    @Override
    public boolean isTestConnectionSupported() {
        return true;
    }

    @Override
    public void testConnection(@NotNull BuildTypeIdentity buildTypeOrTemplate, @NotNull VcsRoot root, @NotNull Map<String, String> params) throws PublisherException {
        UgsStatusPublisher.testConnection(params, trustStore());
    }

    @Nullable
    public CommitStatusPublisher createPublisher(@NotNull SBuildType buildType, @NotNull String buildFeatureId, @NotNull Map<String, String> params) {
        return new UgsStatusPublisher(this, buildType, buildFeatureId, myLinks, params, myProblems);
    }

    @NotNull
    public String describeParameters(@NotNull Map<String, String> params) {
        return String.format("Post commit status to %s", getName());
    }

    @Nullable
    public PropertiesProcessor getParametersProcessor(@NotNull BuildTypeIdentity buildTypeOrTemplate) {
        return new PropertiesProcessor() {
            private boolean checkNotEmpty(@NotNull final Map<String, String> properties,
                                          @NotNull final String key,
                                          @NotNull final String message,
                                          @NotNull final Collection<InvalidProperty> res) {
                if (isEmpty(properties, key)) {
                    res.add(new InvalidProperty(key, message));
                    return true;
                }
                return false;
            }

            private boolean isEmpty(@NotNull final Map<String, String> properties,
                                    @NotNull final String key) {
                return StringUtil.isEmptyOrSpaces(properties.get(key));
            }

            @NotNull
            public Collection<InvalidProperty> process(@Nullable final Map<String, String> p) {
                final Collection<InvalidProperty> result = new ArrayList<InvalidProperty>();
                if (p == null) return result;

                checkNotEmpty(p, UgsConstants.SERVER_URL, "URL must be specified", result);
                // TODO: Check Project formatting as a Perforce URL?
                checkNotEmpty(p, UgsConstants.PROJECT, "Project must be specified", result);
                checkNotEmpty(p, UgsConstants.BADGE_NAME, "Badge name must be specified", result);
                return result;
            }
        };
    }

    @Override
    public boolean isPublishingForVcsRoot(final VcsRoot root) {
        return true;
    }

    @Override
    protected Set<Event> getSupportedEvents(final SBuildType buildType, final Map<String, String> params) {
        return isBuildQueuedSupported(buildType, params) ? mySupportedEventsWithQueued : mySupportedEvents;
    }
}

File: src/main/java/no/tjer/teamcity/ugsStatusPublisher/UgsPublisherSettings.kt
/*
 * Copyright 2023 Jørgen Tjernø <jorgen@tjer.no>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package no.tjer.teamcity.ugsStatusPublisher

import jetbrains.buildServer.commitPublisher.*
import jetbrains.buildServer.serverSide.*
import jetbrains.buildServer.serverSide.auth.SecurityContext
import jetbrains.buildServer.util.StringUtil
import jetbrains.buildServer.util.ssl.SSLTrustStoreProvider
import jetbrains.buildServer.vcs.VcsRoot
import jetbrains.buildServer.web.openapi.PluginDescriptor

/**
 * Settings for Unreal Game Sync commit status publisher.
 */
class UgsPublisherSettings(
    descriptor: PluginDescriptor,
    links: WebLinks,
    problems: CommitStatusPublisherProblems,
    private val mySecurityContext: SecurityContext,
    trustStoreProvider: SSLTrustStoreProvider
) : BasePublisherSettings(descriptor, links, problems, trustStoreProvider), CommitStatusPublisherSettings {
    override fun getId(): String {
        return UgsConstants.ID
    }

    override fun getName(): String {
        return "Unreal Game Sync"
    }

    override fun getEditSettingsUrl(): String {
        return myDescriptor.getPluginResourcesPath("ugsSettings.jsp")
    }

    override fun isTestConnectionSupported(): Boolean {
        return true
    }

    @Throws(PublisherException::class)
    override fun testConnection(buildTypeOrTemplate: BuildTypeIdentity, root: VcsRoot, params: Map<String, String>) {
        UgsStatusPublisher.testConnection(params, trustStore())
    }

    override fun createPublisher(
        buildType: SBuildType, buildFeatureId: String, params: Map<String, String>
    ): CommitStatusPublisher {
        return UgsStatusPublisher(this, buildType, buildFeatureId, myLinks, params, myProblems)
    }

    override fun describeParameters(params: Map<String, String>): String {
        return String.format("Post commit status to %s", name)
    }

    override fun getParametersProcessor(buildTypeOrTemplate: BuildTypeIdentity): PropertiesProcessor {
        return object : PropertiesProcessor {
            private fun checkNotEmpty(
                properties: Map<String, String>, key: String, message: String, res: MutableCollection<InvalidProperty>
            ): Boolean {
                if (isEmpty(properties, key)) {
                    res.add(InvalidProperty(key, message))
                    return true
                }
                return false
            }

            private fun isEmpty(
                properties: Map<String, String>, key: String
            ): Boolean {
                return StringUtil.isEmptyOrSpaces(properties[key])
            }

            override fun process(p: Map<String, String>?): Collection<InvalidProperty> {
                val result: MutableCollection<InvalidProperty> = ArrayList()
                if (p == null) return result
                checkNotEmpty(p, UgsConstants.SERVER_URL, "URL must be specified", result)
                // TODO: Check Project formatting as a Perforce URL?
                checkNotEmpty(p, UgsConstants.PROJECT, "Project must be specified", result)
                checkNotEmpty(p, UgsConstants.BADGE_NAME, "Badge name must be specified", result)
                return result
            }
        }
    }

    override fun isPublishingForVcsRoot(root: VcsRoot): Boolean {
        return true
    }

    override fun getSupportedEvents(
        buildType: SBuildType, params: Map<String, String>
    ): Set<CommitStatusPublisher.Event> {
        return if (isBuildQueuedSupported(buildType, params)) mySupportedEventsWithQueued else mySupportedEvents
    }

    companion object {
        private val mySupportedEvents: Set<CommitStatusPublisher.Event> = setOf(
            CommitStatusPublisher.Event.STARTED,
            CommitStatusPublisher.Event.FINISHED,
            CommitStatusPublisher.Event.INTERRUPTED,
            CommitStatusPublisher.Event.MARKED_AS_SUCCESSFUL
        )
        private val mySupportedEventsWithQueued: Set<CommitStatusPublisher.Event> = mySupportedEvents + setOf(
            CommitStatusPublisher.Event.QUEUED, CommitStatusPublisher.Event.REMOVED_FROM_QUEUE
        )
    }
}
